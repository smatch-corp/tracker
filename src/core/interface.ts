import './global-types.js';

export type TrackerProviderName = keyof TrackerTypes.TrackerProviders;

export interface Options {
  includes?: Partial<Record<TrackerProviderName, true>>;
  excludes?: Partial<Record<TrackerProviderName, true>>;
}

export interface TrackOptions<
  EventProperties extends Record<string, any> = {},
  SessionProperties extends Record<string, any> = {},
> extends Options, TrackerTypes.TrackOptions<EventProperties, SessionProperties> {
  properties?: EventProperties;
  sessionProperties?: SessionProperties;
}

export interface IdentifyOptions extends Options, TrackerTypes.IdentifyOptions {
}

export interface UpdateUserPropertiesOptions<
  UserProperties extends Record<string, any> = {},
> extends Options, TrackerTypes.UpdateUserPropertiesOptions<UserProperties> {
}

export interface IProvider {
  name: string;

  init?: () => void | Promise<void>;

  onTrack?: <
    EventProperties extends Record<string, any> = {},
    SessionProperties extends Record<string, any> = {},
  >(
    eventName: string,
    eventProperties: EventProperties & SessionProperties,
    options: TrackOptions<EventProperties, SessionProperties>,
    context: TrackerTypes.TrackerContext,
  ) => void;

  onIdentify?: (
    id: string,
    options: IdentifyOptions,
    context: TrackerTypes.TrackerContext,
  ) => void;

  onUpdateUserProperties?: <
    UserProperties extends Record<string, any> = {},
  >(
    userProperties: UserProperties,
    options: UpdateUserPropertiesOptions<UserProperties>,
    context: TrackerTypes.TrackerContext,
  ) => void;
}

export interface ITracker {
  track: <
    EventName extends string = string,
    EventProperties extends Record<string, any> = {},
    SessionProperties extends Record<string, any> = {},
  >(eventName: EventName, options?: TrackOptions<EventProperties, SessionProperties>) => void;

  identify: (id: string, options?: IdentifyOptions) => void;

  updateUserProperties: <
    UserProperties extends Record<string, any> = {},
  >(userProperties: UserProperties, options?: UpdateUserPropertiesOptions<UserProperties>) => void;

  clearSessionProperties: () => void;
}
