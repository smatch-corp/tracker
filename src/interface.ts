import './types.js';

export interface TrackerContext extends TrackerTypes.TrackerContext {
}

export type TrackerProviderName = keyof TrackerTypes.TrackerProviders;

export interface TrackOptions<
  EventProperties extends Record<string, any> = {},
  SessionProperties extends Record<string, any> = {},
> extends TrackerTypes.TrackOptions<EventProperties, SessionProperties> {
  properties?: EventProperties;
  sessionProperties?: SessionProperties;
  includes?: Partial<Record<TrackerProviderName, true>>;
  excludes?: Partial<Record<TrackerProviderName, true>>;
}

export interface IProvider {
  name: string;

  init(): void | Promise<void>;

  onTrack: <
    EventProperties extends Record<string, any> = {},
    SessionProperties extends Record<string, any> = {},
  >(
    eventName: string,
    eventProperties: EventProperties & SessionProperties,
    trackOptions: TrackOptions<EventProperties, SessionProperties>,
    context: TrackerContext,
  ) => void;

  onMerge: () => void;

  onIdentify: () => void;

  onUpdateUserProperties: () => void;
}

export interface ITracker {
  track: <
    EventName extends string = string,
    EventProperties extends Record<string, any> = {},
    SessionProperties extends Record<string, any> = {},
  >(eventName: EventName, trackOptions: TrackOptions<EventProperties, SessionProperties>) => void;
}
