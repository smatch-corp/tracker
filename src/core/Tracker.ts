import { IProvider, ITracker, TrackerProviderName, TrackOptions } from '../interface.js';

export interface TrackerOptions {
  providers: IProvider[];
}

export class Tracker implements ITracker {
  #providers: Map<string, IProvider>;
  #initialized: Promise<void>;

  constructor(options: TrackerOptions) {
    this.#providers = new Map();
    for (const provider of options.providers) {
      if (this.#providers.has(provider.name)) {
        throw new Error(`Provider name "${provider.name}" is already in use. Provider names must be unique.`);
      }

      this.#providers.set(provider.name, provider);
    }

    this.#initialized = this.init();
  }

  private async init() {
    const initPromises: Promise<unknown>[] = [];

    for (const provider of this.#providers.values()) {
      const initPromise = Promise.race([
        provider.init(),
        new Promise((_resolve, reject) =>
          setTimeout(reject, 3000, `Provider ${provider.name} is timeout during initalize.`)
        ),
      ]).catch((rejected) => {
        const message: string = typeof rejected === 'string'
          ? `Reason: ${rejected}`
          : rejected instanceof Error
          ? `Caused by:\n  ${rejected.stack}`
          : `Unknown: ${rejected}`;

        console.error(`Provider ${provider.name} is not initialized. ${message}`);

        this.#providers.delete(provider.name);
      });

      initPromises.push(initPromise);
    }

    await Promise.all(initPromises);
  }

  private filterProviders(trackOptions: TrackOptions<any, any>) {
    if (trackOptions.includes && trackOptions.excludes) {
      throw new Error('Cannot set both "includes" and "excludes" options. Choose one or none.');
    }

    if (trackOptions.includes) {
      return Array.from(this.#providers.values())
        .filter(provider => !!trackOptions.includes![provider.name as TrackerProviderName]);
    }

    if (trackOptions.excludes) {
      return Array.from(this.#providers.values())
        .filter(provider => !trackOptions.excludes![provider.name as TrackerProviderName]);
    }

    return Array.from(this.#providers.values());
  }

  track = <
    EventName extends string = string,
    EventProperties extends Record<string, any> = {},
    SessionProperties extends Record<string, any> = {},
  >(eventName: EventName, trackOptions: TrackOptions<EventProperties, SessionProperties>) => {
    this.#initialized.then(() => {
      const trackers = this.filterProviders(trackOptions);

      const eventProperties = {
        ...trackOptions.sessionProperties,
        ...trackOptions.properties,
      } as EventProperties & SessionProperties;

      trackers.forEach(provider => {
        provider.onTrack(eventName, eventProperties, trackOptions, {});
      });
    });
  };

  onMerge(): void {
    throw new Error('Method not implemented.');
  }

  onIdentify(): void {
    throw new Error('Method not implemented.');
  }

  onUpdateUserProperties(): void {
    throw new Error('Method not implemented.');
  }
}
