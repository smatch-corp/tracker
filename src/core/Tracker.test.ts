import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { mock } from 'vitest-mock-extended';
import { IProvider } from './interface.js';
import { Tracker } from './Tracker.js';

const fooProviderName = 'foo' as const;
const barProviderName = 'bar' as const;

declare global {
  export namespace TrackerTypes {
    export interface TrackerProviders {
      [fooProviderName]: true;
      [barProviderName]: true;
    }
  }
}

beforeEach(() => {
  vi.useFakeTimers();
});

afterEach(() => {
  vi.useRealTimers();
});

const setup = () => {
  const provider = mock<IProvider>({
    name: 'foo',
  });

  const createTracker = (providers: IProvider[] = [provider]) => {
    return new Tracker({ providers });
  };

  return { createTracker, provider };
};

describe('init', () => {
  it("should call every providers' isReady methods", () => {
    const { createTracker, provider } = setup();

    createTracker();

    expect(provider.init).toHaveBeenCalled();
  });

  it('should throw error if provider name is not unique', () => {
    const { createTracker, provider } = setup();

    const duplicatedProvider = mock<IProvider>({
      name: provider.name,
    });

    expect(() => createTracker([provider, duplicatedProvider])).toThrowErrorMatchingInlineSnapshot(
      '"Provider name \\"foo\\" is already in use. Provider names must be unique."',
    );
  });

  it('should remove provider if isReady timeout (3s)', async () => {
    const { createTracker, provider } = setup();

    provider.init.mockImplementation(
      () => new Promise(resolve => setTimeout(resolve, 3001)),
    );

    const tracker = createTracker();
    await vi.advanceTimersByTimeAsync(3000);

    tracker.track('test', {
      properties: { foo: 'bar' },
    });
    await vi.runOnlyPendingTimersAsync();

    expect(provider.onTrack).toHaveBeenCalledTimes(0);
  });

  it('should keep provider if isReady resolved in 3s', async () => {
    const { createTracker, provider } = setup();

    provider.init.mockImplementation(
      () => new Promise(resolve => setTimeout(resolve, 2999)),
    );

    const tracker = createTracker();
    await vi.advanceTimersByTimeAsync(3000);

    tracker.track('test', {
      properties: { foo: 'bar' },
    });
    await vi.runOnlyPendingTimersAsync();

    expect(provider.onTrack).toHaveBeenCalledTimes(1);
    expect(provider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {
          "foo": "bar",
        },
        {
          "properties": {
            "foo": "bar",
          },
        },
        {},
      ]
    `);
  });

  it('should remove provider if isReady throw error', async () => {
    const { createTracker, provider } = setup();

    provider.init.mockRejectedValue(new Error('Unknown Error.'));

    const tracker = createTracker();
    tracker.track('test', {
      properties: { foo: 'bar' },
    });
    await vi.runOnlyPendingTimersAsync();

    expect(provider.onTrack).toHaveBeenCalledTimes(0);
  });

  it("should call providers' isReady once", async () => {
    const { createTracker, provider } = setup();

    const tracker = createTracker();
    tracker.track('test', {});
    tracker.track('test', {});
    tracker.track('test', {});

    await vi.runAllTimersAsync();

    expect(provider.init).toHaveBeenCalledTimes(1);
    expect(provider.onTrack).toHaveBeenCalledTimes(3);
  });
});

describe('track', () => {
  it("should call every providers' track methods", async () => {
    const { createTracker, provider } = setup();
    const yetAnotherProvider = mock<IProvider>({
      name: 'yetAnotherProvider',
    });

    const tracker = createTracker([provider, yetAnotherProvider]);

    tracker.track('test', {
      properties: { foo: 'bar' },
    });

    await vi.runAllTimersAsync();

    expect(provider.onTrack).toHaveBeenCalledTimes(1);
    expect(provider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {
          "foo": "bar",
        },
        {
          "properties": {
            "foo": "bar",
          },
        },
        {},
      ]
    `);

    expect(yetAnotherProvider.onTrack).toHaveBeenCalledTimes(1);
    expect(yetAnotherProvider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {
          "foo": "bar",
        },
        {
          "properties": {
            "foo": "bar",
          },
        },
        {},
      ]
    `);
  });

  it('should not call track method of provider which not specified in trackerOptions.includes', async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);
    tracker.track('test', {
      properties: { foo: 'bar' },
      includes: { bar: true },
    });

    await vi.runAllTimersAsync();

    expect(fooProvider.onTrack).toHaveBeenCalledTimes(0);

    expect(barProvider.onTrack).toHaveBeenCalledTimes(1);
    expect(barProvider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {
          "foo": "bar",
        },
        {
          "includes": {
            "bar": true,
          },
          "properties": {
            "foo": "bar",
          },
        },
        {},
      ]
    `);
  });

  it('should not call track method of provider which specified in trackerOptions.excludes', async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);
    tracker.track('test', {
      properties: { foo: 'bar' },
      excludes: { bar: true },
    });

    await vi.runAllTimersAsync();

    expect(fooProvider.onTrack).toHaveBeenCalledTimes(1);
    expect(fooProvider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {
          "foo": "bar",
        },
        {
          "excludes": {
            "bar": true,
          },
          "properties": {
            "foo": "bar",
          },
        },
        {},
      ]
    `);

    expect(barProvider.onTrack).toHaveBeenCalledTimes(0);
    expect(barProvider.onTrack.mock.lastCall).toMatchInlineSnapshot('undefined');
  });

  describe('sessionProperties', () => {
    it('should store sessionProperties in tracker and merge properties', async () => {
      const { createTracker, provider } = setup();

      const tracker = createTracker();

      tracker.track('test', {
        properties: { productId: '1' },
        sessionProperties: { userId: '2' },
      });

      await vi.runAllTimersAsync();

      expect(provider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
        [
          "test",
          {
            "productId": "1",
            "userId": "2",
          },
          {
            "properties": {
              "productId": "1",
            },
            "sessionProperties": {
              "userId": "2",
            },
          },
          {},
        ]
      `);

      tracker.track('test', {
        properties: { productId: '2' },
      });

      await vi.runAllTimersAsync();

      expect(provider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
        [
          "test",
          {
            "productId": "2",
            "userId": "2",
          },
          {
            "properties": {
              "productId": "2",
            },
          },
          {},
        ]
      `);
    });

    it('should clear sessionProperties after call clearSessionProperties', async () => {
      const { createTracker, provider } = setup();

      const tracker = createTracker();

      tracker.track('test', {
        properties: { productId: '1' },
        sessionProperties: { userId: '2' },
      });

      await vi.runAllTimersAsync();

      expect(provider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
        [
          "test",
          {
            "productId": "1",
            "userId": "2",
          },
          {
            "properties": {
              "productId": "1",
            },
            "sessionProperties": {
              "userId": "2",
            },
          },
          {},
        ]
      `);

      tracker.clearSessionProperties();

      tracker.track('test', {
        properties: { productId: '2' },
      });

      await vi.runAllTimersAsync();

      expect(provider.onTrack.mock.lastCall).toMatchInlineSnapshot(`
        [
          "test",
          {
            "productId": "2",
          },
          {
            "properties": {
              "productId": "2",
            },
          },
          {},
        ]
      `);
    });
  });
});

describe('identify', () => {
  it("should call every providers' identify methods", async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);

    tracker.identify('test');

    await vi.runAllTimersAsync();

    expect(fooProvider.onIdentify).toHaveBeenCalledTimes(1);
    expect(fooProvider.onIdentify.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {},
        {},
      ]
    `);

    expect(barProvider.onIdentify).toHaveBeenCalledTimes(1);
    expect(barProvider.onIdentify.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {},
        {},
      ]
    `);
  });

  it('should not call identify method of provider which not specified in options.includes', async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);
    tracker.identify('test', {
      includes: { bar: true },
    });

    await vi.runAllTimersAsync();

    expect(fooProvider.onIdentify).toHaveBeenCalledTimes(0);

    expect(barProvider.onIdentify).toHaveBeenCalledTimes(1);
    expect(barProvider.onIdentify.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {
          "includes": {
            "bar": true,
          },
        },
        {},
      ]
    `);
  });

  it('should not call identify method of provider which specified in options.excludes', async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);
    tracker.identify('test', {
      excludes: { bar: true },
    });

    await vi.runAllTimersAsync();

    expect(fooProvider.onIdentify).toHaveBeenCalledTimes(1);
    expect(fooProvider.onIdentify.mock.lastCall).toMatchInlineSnapshot(`
      [
        "test",
        {
          "excludes": {
            "bar": true,
          },
        },
        {},
      ]
    `);

    expect(barProvider.onIdentify).toHaveBeenCalledTimes(0);
    expect(barProvider.onIdentify.mock.lastCall).toMatchInlineSnapshot('undefined');
  });
});

describe('updateUserProperties', () => {
  it('should call every providers` updateUserProperties methods', async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);
    tracker.updateUserProperties({ name: 'John' });

    await vi.runAllTimersAsync();

    expect(fooProvider.onUpdateUserProperties).toHaveBeenCalledTimes(1);
    expect(fooProvider.onUpdateUserProperties.mock.lastCall).toMatchInlineSnapshot(`
      [
        {
          "name": "John",
        },
        {},
        {},
      ]
    `);

    expect(barProvider.onUpdateUserProperties).toHaveBeenCalledTimes(1);
    expect(barProvider.onUpdateUserProperties.mock.lastCall).toMatchInlineSnapshot(`
      [
        {
          "name": "John",
        },
        {},
        {},
      ]
    `);
  });

  it('should not call updateUserProperties method of provider which not specified in options.includes', async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);
    tracker.updateUserProperties({ name: 'John' }, {
      includes: { bar: true },
    });

    await vi.runAllTimersAsync();

    expect(fooProvider.onUpdateUserProperties).toHaveBeenCalledTimes(0);

    expect(barProvider.onUpdateUserProperties).toHaveBeenCalledTimes(1);
    expect(barProvider.onUpdateUserProperties.mock.lastCall).toMatchInlineSnapshot(`
      [
        {
          "name": "John",
        },
        {
          "includes": {
            "bar": true,
          },
        },
        {},
      ]
    `);
  });

  it('should not call updateUserProperties method of provider which specified in options.excludes', async () => {
    const { createTracker, provider: fooProvider } = setup();
    const barProvider = mock<IProvider>({
      name: 'bar',
    });

    const tracker = createTracker([fooProvider, barProvider]);
    tracker.updateUserProperties({ name: 'John' }, {
      excludes: { bar: true },
    });

    await vi.runAllTimersAsync();

    expect(fooProvider.onUpdateUserProperties).toHaveBeenCalledTimes(1);
    expect(fooProvider.onUpdateUserProperties.mock.lastCall).toMatchInlineSnapshot(`
      [
        {
          "name": "John",
        },
        {
          "excludes": {
            "bar": true,
          },
        },
        {},
      ]
    `);

    expect(barProvider.onUpdateUserProperties).toHaveBeenCalledTimes(0);
    expect(barProvider.onUpdateUserProperties.mock.lastCall).toMatchInlineSnapshot('undefined');
  });
});
