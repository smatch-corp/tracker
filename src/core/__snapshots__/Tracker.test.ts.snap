// Vitest Snapshot v1

exports[`init > should keep provider if isReady resolved in 3s 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`init > should keep provider if isReady resolved in 3s 2`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`init > should remove provider if isReady throw error 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`init > should remove provider if isReady timeout (3s) 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`init > should remove provider if isReady timeout (3s) 2`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`init > should throw error if provider name is not unique 1`] = `"Provider name \\"bar\\" is already in use. Provider names must be unique."`;

exports[`track > should call every providers' track methods 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`track > should call every providers' track methods 2`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`track > should not call track method of provider which not specified in trackerOptions.includes 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "test",
      {
        "foo": "bar",
      },
      {
        "includes": {
          "bar": true,
        },
        "properties": {
          "foo": "bar",
        },
      },
      {},
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
